prereise/gather/demanddata/transportation_electrification/smart_charging.py

```diff
@@ -117,6 +123,90 @@ def smart_charging(
     else:
         charging_efficiency = 0.9
 
+    # Add booleans for whether the location allows charging
+    if location_strategy == 3:
+        newdata["location_allowed"] = True
+    else:
+        if veh_type.lower() in {"ldv", "ldt"}:
+            allowed = const.ldv_location_allowed[location_strategy]
+            newdata["location_allowed"] = newdata["dwell_location"].isin(allowed)
+        elif veh_type.lower() in {"mdv", "hdv"}:
+            allowed = const.hdv_location_allowed[location_strategy]
+            newdata["location_allowed"] = newdata["dwell_location"].isin(allowed)
+
+    # Add booleans for whether the trip_number (compared to total trips) allows charging
+    if trip_strategy == 1:
+        newdata["trip_allowed"] = True
+    elif trip_strategy == 2:
+        newdata["trip_allowed"] = newdata["trip_number"] == newdata["total_trips"]
+
+    # Add booleans for whether the dwell time is long enough to allow charging
+    newdata["dwell_allowed"] = newdata["dwell_time"] > 0.2
+
+    # Add boolean for whether this trip allows charging
+    allowed_cols = [
+        "location_allowed",
+        "trip_allowed",
+        "dwell_allowed",
+    ]
+    newdata["charging_allowed"] = newdata[allowed_cols].apply(all, axis=1)
+
+    newdata["dwell_charging"] = (
+        newdata["charging_allowed"]
+        * newdata["dwell_time"]
+        * power
+        * charging_efficiency
+    )
+
+    grouped_trips = newdata.groupby("vehicle_number")
+    for vehicle_num, group in grouped_trips:
+        newdata.loc[group.index, "max_charging"] = newdata.loc[
+            group.index, "dwell_charging"
+        ].sum()
+        newdata.loc[group.index, "required_charging"] = (
+            newdata.loc[group.index, "trip_miles"].sum() * kwhmi
+        )
+
+    # Filter for whenever available charging is insufficient to meet required charging
+    newdata = newdata.loc[
+        (newdata["required_charging"] <= newdata["max_charging"])
+    ].copy()
+
+    # Filter for vehicle's battery range
+    newdata = newdata.loc[(newdata["total vehicle miles traveled"] < veh_range)]
+
+    nd_len = len(newdata)
+
+    # LDV and LDT filter for cyclical trips
+    if veh_type.lower() in {"ldv", "ldt"}:
+        filtered_census_data = pd.DataFrame(columns=const.nhts_census_column_names)
+        i = 0
+        while i < nd_len:
+            total_trips = int(newdata.iloc[i, newdata.columns.get_loc("total_trips")])
+            # copy one vehicle information to the block
+            individual = newdata.iloc[i : i + total_trips].copy()
+            if (
+                individual["why_from"].head(1).values[0]
+                == individual["dwell_location"].tail(1).values[0]
+            ):
+                filtered_census_data = pd.concat(
+                    [filtered_census_data, individual], ignore_index=True
+                )
+            i += total_trips
+        del newdata
+        gc.collect()
```

Comment:
Just curious, are we on a memory sensitive situation so that we would like to release the memory as soon as possible? Also we didn't call the garbage collector in the same logic during immediate charging above. Why in that case we simply want to remove the reference rather than reclaim the memory?

Reply:
I think we can take this out, I was concerned that an error I was seeing earlier might be memory-related